### ISSUE TITLE
Refactor the code to make it compliant with Python 3 typing

### ISSUE THREAD

**Issue by nicolargo** [nicolargo/glances#2404]
Glances has been developed for Python 2. The next version (4.0.0) will only be compliant with Python 3.8 or higher.

It's time to refactor the code to provides runtime support for type hints: https://docs.python.org/3/library/typing.html

@RazCrimson any experiences/tips to convert a Python code for Python 3 typing ? Do you think that we must include an external type checker (like mypi or others) ?

Some resources found on the net:

* https://towardsdatascience.com/modernize-your-sinful-python-code-with-beautiful-type-hints-4e72e98f6bf1
* https://realpython.com/python-type-checking/
* https://docs.python.org/3/library/typing.html

**Comment by nicolargo** [comment]
Also related: 

- https://github.com/nicolargo/glances/issues/2182

**Comment by RazCrimson** [comment]
@nicolargo

> @RazCrimson any experiences/tips to convert a Python code for Python 3 typing ? Do you think that we must include an external type checker (like mypi or others)

Yes, mypy would be nice to add it

For glances but since most of the our variables are dicts, typing them directly would not be really useful.

We should start typing after we define the models (pydantic issue mentioned above), otherwise we would have the change the types of all those dicts into their corresponding pydantic equivalents later which would be extra work.

**Comment by nicolargo** [comment]
Hi @RazCrimson on the https://github.com/nicolargo/glances/tree/2404-refactor-the-code-to-make-it-compliant-with-python-3-typing branch, i start the refactoring by creating two news dataclasses: GlancesDataItem and GlancesDataPlugin. 

Those classes are for the moment only used in the CPU and MEM plugins and simplify the plugins code.

Can you have a first look and tell me if you see others thinks to improve and will simplify the API documentation. 

For the moment, typing is limited to the new dataclasses and there is no verification (non pydantic for the moment).

**Comment by RazCrimson** [comment]
@nicolargo 

Sure, will take a look at them over the weekend!

**Comment by nicolargo** [comment]
The current implementation (june 10, 2023) only take "simple" plugin into account. By "simple" i mean plugin which return a simple dictionary (like cpu, mem, ...):

```
{"total":7836184576,"available":1656225792,"percent":78.9,"used":11868876800,"free":4032692224,"active":3123007488,"inactive":3336036352,"buffers":200724480,"cached":2175741952,"wired":null,"shared":653058048}
```

but not for plugins returning list of dict (like percpu, network, fs...):

```
[{"device_name":"\/dev\/mapper\/ubuntu--gnome--vg-root","fs_type":"ext4","mnt_point":"\/","size":243334156288,"used":201482113024,"free":29464588288,"percent":87.2,"key":"mnt_point"},{"device_name":"zsfpool","fs_type":"zfs","mnt_point":"\/zsfpool","size":41943040,"used":131072,"free":41811968,"percent":0.3,"key":"mnt_point"},{"device_name":"\/dev\/mapper\/ubuntu--gnome--vg-root","fs_type":"ext4","mnt_point":"\/var\/snap\/firefox\/common\/host-hunspell","size":243334156288,"used":201482113024,"free":29464588288,"percent":87.2,"key":"mnt_point"}]
```

I see 3 solutions to refactor the code for those plugins:

1) Use fields_description = [ { 'item_key': item_value... } ]
The definition is done done in the first dict. One of the item_key define the key name (ex: mnt_point for the FS)
Pro: no change in the data model (regarding Glances version 3)
Con: do not know how to adapt the GlancesDataPlugin class for this kind of data model. We should probably make some test like "if fields_description is list"...

2) Use fields_description =  { 'key': { 'item_key': item_value...}, ...}

Pro: The fields_description is always a dict. But item value can be dict 
Con:

**Comment by nicolargo** [comment]
Additional question: How to deal with plugin returning list of dict ?

In this case, how to define the plugin with fields_description ?

The data model (after export) could:
- be the same than Glances version 3: a list of dict
- change: a dict of dict
